### PR TITLE
Proof of Concept: Multiple windows [#5816]

### PR DESCRIPTION
**Pull request by junbio**
Here is an experimental build that implements multiple window support (#5816). It's still in a very early state, so I don't expect it to go into production anytime soon, if at all. The following functionality work:

- Dragging a tab between windows
- Dragging a tab into its own window
- Long press menu to open link in a new window
- Windows are saved between sessions

The main problem is that much of the code relies on the singleton BrowserViewController.foregroundBVC() such as the new settings view model classes. I attempted to solve this by adding a weak reference to the browser view controller in TabManager, but this might not be ideal. I also had to change the file format of how TabManagerStore stores tab data on disk in order to support multiple windows.

Since this is such a fundamental breaking change to how Firefox works, I understand if Mozilla team members would like to tackle this on their own. I hope I've been helpful but I apologize for any inconvenience.

### CODE REVIEW

**Comment by nishant2718** [comment]
This is an awesome PoC @junbio ! We've hit some of the same issues you mentioned, and we're working out what the best approach to support multi windows would be step by step. There's some prerequisite work that I think needs to happen to support multiple windows well, and you're definitely heading the right direction here. 

I'll be looking over this PR and check in with you again soon. Thanks again for the great work!

**Comment by junbio** [comment]
Thanks for the feedback @nishant2718!

>There's some prerequisite work that I think needs to happen to support multiple windows well, and you're definitely heading the right direction here.

One of the issues that I encountered is that the theme change is done in a singleton BrowserViewController traitCollectionDidChange method, which isn't true when there are multiple windows. Also, iOS takes several snapshots of each window in both light and dark mode, which causes flickering. One of the items on the roadmap I saw was migrating to .xcasset from the old theming system, which won't scale and isn't needed anymore since the minimum iOS version is 13. I actually am working on this migration if you're interested.

**Comment by nishant2718** [comment]
That migration's much needed - it's actually an item on @electricRGB 's list. He has an idea of how to go about it, and I'll let him know you're also interested in that area. If you're not on it already, I recommend you join chat.mozilla.org, and the Firefox-iOS channel. Let us know when you get there :)

**Comment by junbio** [comment]
So I went ahead and opened another PR with a preliminary migration to .xcassets #11343. I hope you don't mind. Thanks!

File: Client/Application/AppDelegate.swift [hunk 1]
```diff
@@ -271,4 +272,28 @@ extension AppDelegate {
 
         window!.rootViewController = rootViewController
     }
+    
+@available(iOS 13.0, *)
```
[the remark this answers is not in the document]
Comment (junbio) [reply Client/Application/AppDelegate.swift:276]: Thanks for the feedback! Yeah I should have cleaned up better as I started work on this two years ago.

**Comment by nishant2718** [comment]
hey @junbio ! It's been a while and I should update you on the state of things. 

First of all, this is amazing work - you've implemented features we're building towards! We always appreciate your contributions and look forward to seeing them. The team took a deep dive into the PR but we had to take a step back to see the bigger picture. 

The transition to Scenes is a risky task. The reason we’re taking it slow is because we’ve identified quite a few things that needed to happen leading up to being able to work on this, like:
- IDing our initial blockers, like the usage of `foregroundBVC()`
- handling our dependencies reasonably (we’re slowly adopting a container based strategy with [Dip](https://github.com/AliSoftware/Dip))
- Isolating window / view setup that’s within AppDelegate
- considerations of resource usage / management with multi window enabled. Will we ever face race conditions or deadlocks?
- much more

We try giving context where we can, but sometimes it’s a bit difficult to convey. And that’s why it took me a while to respond. I hope this helps in understanding the journey to a Scened app and eventually, multiple window support :)

Again, we really appreciate the incredible work in this PR and all your contributions! [The PR we have](https://github.com/mozilla-mobile/firefox-ios/pull/12241) up now accomplishes a portion of you've done here - we hope to take small steps towards a successful and stable multi window app. I'll list you as a co-author there as well! 

Hope to see you in the next ones 😄

**Pull request closed without merging**